### DISCUSSION TITLE
Error handling issues and inconsistencies

### DISCUSSION THREAD

**Discussion by tommie** [nuxt/nuxt#23679] [General]
I've had a feeling that I don't understand Nuxt's error handling, so I made [this playground](https://stackblitz.com/edit/github-whclvz-jrpsbr?file=app.vue). It can trigger errors using `createError`, 404 and 500 in various phases.

I'm not yet at a point where I could make recommendations, and I may have screwed something up in the playground, but I'll post this for discussion.

## Summary

* `useError` never seems to be set to anything useful. On the error page, we use the `error` prop, and in other pages, it's never set. It's surprising to me that Nuxt doesn't internally set this based on hooks/callbacks. I'd have to emulate it at the error boundary.
* `<NuxtLink to="/" @click.prevent=clearError>` and `<button @click=clearError>` don't work the same when using redirect.
* That the error page shows up *sometimes*, even within an error boundary, means e.g. showing toasts for errors is highly inconsistent. I can understand that 404 triggers the error page (though isn't that really a result of `NuxtPage` failing?), but 500 inside `onServerPrefetch` seems a stretch.

(It seems that StackBlitz doesn't update the URL on navigation, though that doesn't seem to be the issue. The root page and `/200` do the same thing. I just wanted to see if `clearError('/')` was different on `/200` or not.)

## `clearError` doesn't clear thrown `createError`

At the root page, if I click `createError`, the root page is unmounted, and error hooks are called. `useError` is undefined. Neither of the `clearError` buttons do anything useful. `reloadNuxtApp` gets me back to where I'd expect `clearError` to land me. Since I suspected `redirect:'/'` is optimized away at the root page, I created `/200`. But it has the same behavior: nothing happens.

## Error hooks for 404 are different, depending on how I got to the 404

1. At the root page, click `a 404`. This brings you to the error page with no error hooks run.
2. Click the `clearError redirect` button. Back to the root page with only error:cleared hooks run. Using the `nuxt-link.prevent clearError redirect` does the same.
3. At the root page, click `nuxt-link 404`. This brings you to the error page with error hooks run.
4. Click the `clearError redirect` button. Back to the root page with error hooks run. Using the `nuxt-link.prevent clearError redirect` does the same.

The explanation is (probably) that the error hooks ran before the `<a>` navigated away from the page, but it means I need to remember to clear these errors for `nuxt-links`, even though the error page was used. It would make more sense if the `app:error`s were triggered after navigation.

## `clearError` does not handle follow-on errors the same way as navigation errors

1. At the root page, click `a 500`. This brings you to the error page.
2. Click the `clearError` button or `nuxt-link clearError`. Either will get you back to the `/500`. The page isn't rendered, because of the error boundary.
3. At the root page, click `nuxt-link 500`. This triggers the error boundary, not the error page.

The first time, the error happens on navigation, and using `clearError` or `nuxt-link` is a vue-router route change. It would be much more obvious if the error boundary was also honored on initial navigation.

## `nuxt-link` with `@click.prevent` is not the same as `button @click`

1. At the root page, click `a 500`. This brings you to the error page.
2. Click the `clearError redirect` button. You are brought back to the root page.
3. At the root page, click `a 500`. This brings you to the error page.
4. Click the `nuxt-link.prevent redirect` lnik. You are still at the `/500` page, with the `on mounted` error (no page is loaded.)
5. At the root page, click `nuxt-link 500`. This brings you to the error page.
6. Click the `clearError redirect` button. You are brought back to the root page.
7. Now this button that worked before, has the same effect as `nuxt-link.prevent redirect` at step 4.

Perhaps `.prevent` isn't enough to force using `clearError`, but the inconsistency between (4) and (7) are worrisome.

## Note on the `500 on setup()`

It seems these are consistently working the same as `500`, which is good.

**Comment by 2113ic** [comment]
To my surprise, we're almost in 2026 and this issue still exists, with no discussion around it at all.
